cmd/gcp-controller-manager/app/csrmetrics/csrmetrics.go

```diff
@@ -0,0 +1,121 @@
+)
+
+// CSRSigningStatus is a status string of the CSR Signing metric.
+type CSRSigningStatus string
+
+// CSRApprovalStatus is a status string of the CSR Approval metric.
+type CSRApprovalStatus string
+
+// OutboundRPCStatus is a status string of the Outbound RPC metric.
+type OutboundRPCStatus string
+
+// CSRSigningRecord is a function to record the CSR Signing metric with a status.
+type CSRSigningRecord func(status CSRSigningStatus)
+
+// CSRApprovalRecord is a function to record the CSR Approval metric with a status.
+type CSRApprovalRecord func(status CSRApprovalStatus)
+
+// OutboundRPCRecord is a function to record the Outbound RPC metric with a status.
+type OutboundRPCRecord func(status OutboundRPCStatus)
+
+// Status constants for metrics.
+const (
+	CSRSigningStatusSignError   CSRSigningStatus = "sign_error"
+	CSRSigningStatusUpdateError CSRSigningStatus = "update_error"
+	CSRSigningStatusSigned      CSRSigningStatus = "signed"
+
+	CSRApprovalStatusNodeDeleted         CSRApprovalStatus = "node_deleted"
+	CSRApprovalStatusParseError          CSRApprovalStatus = "parse_error"
+	CSRApprovalStatusSARError            CSRApprovalStatus = "sar_error"
+	CSRApprovalStatusSARReject           CSRApprovalStatus = "sar_reject"
+	CSRApprovalStatusPreApproveHookError CSRApprovalStatus = "pre_approve_hook_error"
+	CSRApprovalStatusDeny                CSRApprovalStatus = "deny"
+	CSRApprovalStatusApprove             CSRApprovalStatus = "approve"
+	CSRApprovalStatusIgnore              CSRApprovalStatus = "ignore"
+
+	OutboundRPCStatusNotFound OutboundRPCStatus = "not_found"
+	OutboundRPCStatusError    OutboundRPCStatus = "error"
+	OutboundRPCStatusOK       OutboundRPCStatus = "ok"
+)
+
+var (
+	csrSigningCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "csr_signing_count",
+		Help: "Count of signed CSRs",
+	}, []string{"status"})
+	csrSigningLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "csr_signing_latencies",
+		Help: "Latency of CSR signer, in seconds",
+	}, []string{"status"})
+	csrApprovalCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "csr_approval_count",
+		Help: "Count of approved, denied and ignored CSRs",
+	}, []string{"status", "kind"})
+	csrApprovalLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "csr_approval_latencies",
+		Help: "Latency of CSR approver, in seconds",
+	}, []string{"status", "kind"})
+	outboundRPCLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "outbound_rpc_latency",
+		Help: "Latency of outbound RPCs to GCE and GKE, in seconds",
+	}, []string{"status", "kind"})
+)
+
+func init() {
+	prometheus.MustRegister(csrSigningCount)
+	prometheus.MustRegister(csrSigningLatency)
+	prometheus.MustRegister(csrApprovalCount)
+	prometheus.MustRegister(csrApprovalLatency)
+	prometheus.MustRegister(outboundRPCLatency)
+}
+
+// CSRSigningStartRecorder starts a timer for kind and returns a
+// CSR Singing status record function.
```

Comment:
nit: "SigningStartRecorder marks the start of a CSR signing operation. Caller is responsible for calling the returned function, which records Prometheus metrics for this operation."